packages/rpc-transport/src/json-rpc-subscription.ts

```diff
@@ -0,0 +1,142 @@
+import { JsonRpcResponse } from './json-rpc';
+import { SolanaJsonRpcError } from './json-rpc-errors';
+import { createJsonRpcMessage } from './json-rpc-message';
+import {
+    PendingRpcSubscription,
+    RpcSubscription,
+    RpcSubscriptionConfig,
+    RpcSubscriptions,
+    SubscribeOptions,
+} from './json-rpc-types';
+
+type JsonRpcNotification<TNotification> = Readonly<{
+    params: Readonly<{
+        result: TNotification;
+        subscription: number;
+    }>;
+}>;
+type SubscriptionId = number;
+
+function registerIterableFatal(iterable: AsyncIterable<unknown>, catchFn: CallableFunction) {
+    (async () => {
+        try {
+            // eslint-disable-next-line @typescript-eslint/no-unused-vars
+            for await (const _ of iterable);
+        } catch (e) {
+            // Run the catch function if the iterator throws.
+            catchFn(e);
+        }
+    })();
+}
+
+function createPendingRpcSubscription<TRpcSubscriptionMethods, TNotification>(
+    rpcConfig: RpcSubscriptionConfig<TRpcSubscriptionMethods>,
+    { params, subscribeMethodName, unsubscribeMethodName, responseProcessor }: RpcSubscription<TNotification>
+): PendingRpcSubscription<TNotification> {
+    return {
+        async subscribe(options?: SubscribeOptions): Promise<AsyncIterable<TNotification>> {
+            options?.abortSignal?.throwIfAborted();
+            let subscriptionId: number | undefined;
+            function handleCleanup() {
+                if (subscriptionId !== undefined) {
+                    const payload = createJsonRpcMessage(unsubscribeMethodName, [subscriptionId]);
+                    const unsubscribeAbortController = new AbortController();
+                    rpcConfig.transport({ payload, signal: unsubscribeAbortController.signal }).finally(() => {
+                        unsubscribeAbortController.abort();
+                    });
+                }
+                connectionAbortController.abort();
```

Comment:
Seems like this could just use one abort controller, no? why not just pass `connectionAbortController` into the closure?

Reply:
You're right. I just need to change this to ‘kill the one now if you don't have a subscription’ or ‘kill both once you've sent the unsubscribe message.’